### ISSUE TITLE
github-runner: systemd hardening too restrictive

### ISSUE THREAD

**Issue by veehaitch** [NixOS/nixpkgs#181573]
### Describe the bug

#163030 introduced systemd syscall filtering which seems overly restrictive for common use cases.

### Steps To Reproduce

- We use `cargo-tarpaulin` for coverage reports of our Rust projects which makes use of the _ptrace(2)_ system call (denied by `~@debug`).
- Android projects commonly rely on `arm-linux-androideabi-ar` which uses _chown(2)_ (denied by `~@privileged`).

### Expected behavior

In my opinion, those use cases are too common and should work out of the box. Also, workflow failures are difficult to debug if syscall filtering (or any hardening) steps in.

I'd suggest to revise the syscall filtering:

https://github.com/NixOS/nixpkgs/blob/26839ce8e094340fb89a718b07a4807eebd984f4/nixos/modules/services/continuous-integration/github-runner.nix#L304-L310

Also, it might help with debugging or employing a custom hardening configuration if we add a module option `enableHardening` which defaults to `true`.

### Notify maintainers

@newAM @kfollesdal 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 5.18.6, NixOS, 22.11 (Raccoon), 22.11.20220629.cf3ab54`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.9.1`
 - nixpkgs: `/nix/store/8mik2m5inkjin29v5gv5zc2jji4lab3l-source`
output here
```

**Comment by newAM** [comment]
@aanderse also noted problems with makefile workflows in #163030 that I have been able to reproduce.

I am sorry I did not catch this when making the changes originally.  My usecase was running binaries on uncommon targets (e.g. `thumbv7em-none-eabi`) connected over USB; and I only had light testing for other usecases.  I have since added a makefile test to the runner I use for testing package/module updates and will add a coverage workflow now too.

I am ok with removing the `SystemCallFilter` harderning entirely, or making it more permissive.  I only added this because at the time I thought it would be a sane default for majority of applications.  I would prefer we keep _some_ hardening.  Most people probably do not want a CI process to change the clock, hostname, or reboot the system entirely.

I have been working on this in the background since @aanderse reported issues; this is what I have come up with so far, I have not tested it with `cargo-tarpaulin` yet (will do that later this weekend); but it does allow `make` and debug related tools such as `strace` to function.

```nix
{
  SystemCallFilter = [
    "~@capset"
    "~@clock"
    "~@cpu-emulation"
    "~@module"
    "~@mount"
    "~@obsolete"
    "~@raw-io"
    "~@reboot"
    "~@setdomainname"
    "~@sethostname"
  ];
  # Prevents "GC Warning: Couldn't read /proc/stat" when running nix
  ProcSubset = "all";
}
```

In addition to `SystemCallFilter` changes `ProcSubset` should also be changed to `"all"` so that nix does not emit warnings when running.

**Comment by aanderse** [comment]
> Most people probably do not want a CI process to change the clock, hostname, or reboot the system entirely.

:+1:

> In addition to `SystemCallFilter` changes `ProcSubset` should also be changed to `"all"` so that nix does not emit warnings when running.

:+1:

> we should remove them or make them optional

Yes, all this hardening is already available and entirely optional! Every user is free to add anything to `systemd.services.github-runner.serviceConfig` that they want. I understand that NixOS developers are super keen on `systemd` hardening, which is generally good, but for certain applications I just don't think it makes sense to go over the top... or even provide "curated" selections behind 'enableHardening' type flags. CI software is so broad I think we're providing more trouble and confusion than value here.

My personal recommendation for this specific module is to follow what @newAM suggested as a reasonable default and let users decide any hardening they want for themselves. But I'm not the maintainer of this module in NixOS, just a user, so I defer to those who are volunteering their time to maintain this module :bowing_man:

**Comment by newAM** [comment]
PR to fix is here: #181867

Mentions do not generate notifications, just making a comment so that anyone subscribed to this issue gets a notification.
